crypto/sm2/sm2_pmeth.c

```diff
@@ -22,28 +22,35 @@ typedef struct {
     EC_GROUP *gen_group;
     /* message digest */
     const EVP_MD *md;
+    /* Distinguishing Identifier, ISO/IEC 15946-3 */
+    uint8_t *id;
+    int id_len;
 } SM2_PKEY_CTX;
 
 static int pkey_sm2_init(EVP_PKEY_CTX *ctx)
 {
-    SM2_PKEY_CTX *dctx;
+    SM2_PKEY_CTX *smctx;
 
-    if ((dctx = OPENSSL_zalloc(sizeof(*dctx))) == NULL) {
+    if ((smctx = OPENSSL_zalloc(sizeof(*smctx))) == NULL) {
         SM2err(SM2_F_PKEY_SM2_INIT, ERR_R_MALLOC_FAILURE);
         return 0;
     }
 
-    ctx->data = dctx;
+    /* -1 indicates there is no ID set yet */
+    smctx->id_len = -1;
```

Comment:
...but this is problematic if it were to be `size_t`. I still think it should be though...I had too much pain sorting out the "it's a  size_t, no it's a long, no it's an int" mess that we used to have in libssl (casts everywhere).

Reply:
> Perhaps just add an additional flag into the ctx?

That is workable